src/merkle/tiered_smt/index.rs

```diff
@@ -0,0 +1,405 @@
+
+    /// Computes the index for the current depth.
+    pub const fn index(&self) -> u64 {
+        self.canonical >> (64 - self.depth)
+    }
+
+    /// Returns true if the current instance is at the maximum depth of `[TieredSmt]`.
+    pub const fn is_max_depth(&self) -> bool {
+        self.depth == TieredSmt::MAX_DEPTH
+    }
+    /// Computes the remaining path of a leaf for the current depth.
+    ///
+    /// Will mutate only the last element of the key as it is the only element used to traverse the
+    /// tree.
+    pub fn remaining_path(&self) -> CanonicalWord {
+        self.remaining_path_with_depth(self.depth)
+    }
+
+    /// Computes the remainig path of a leaf for an arbitrary depth.
+    pub fn remaining_path_with_depth(&self, depth: u8) -> CanonicalWord {
+        // clippy incorrectly assumes the function when `None` is a constant/copy, so it is cheap
+        // enough to be executed by default. however, `64` is a valid input for this function, and
+        // it would cause overflow.
+        //
+        // satisfying clippy here isn't worthy as we might greatly increase the complexity of this
+        // function to do so.
+        #[allow(clippy::unnecessary_lazy_evaluations)]
+        let value = (depth == TieredSmt::MAX_DEPTH)
+            .then_some(0)
+            .unwrap_or_else(|| (self.canonical << depth) >> depth);
+        self.key.with_path(value)
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Traverse the leaf to the next tier. Returns `false` if the current tear is the maximum and
+    /// no mutation was performed.
+    pub fn traverse(&mut self) -> bool {
+        if self.depth >= TieredSmt::MAX_DEPTH {
+            self.depth = TieredSmt::MAX_DEPTH + TieredSmt::TIER_DEPTH;
+            return false;
+        }
+        self.depth += TieredSmt::TIER_DEPTH;
+        true
+    }
+
+    /// Traverse the index to the first tier, returning it's correspondent node index.
+    pub fn traverse_to_first_tier(&mut self) -> NodeIndex {
+        self.depth = TieredSmt::TIER_DEPTH;
+        self.into()
+    }
+
+    /// Traverse to a given depth.
+    ///
+    /// This function will not evaluate if the provided depth is greater than the maximum allowed
+    /// for `[TieredSmt]`.
+    pub fn traverse_to(&mut self, depth: u8) -> &mut Self {
+        // normalize the depth into a tier, truncating any intermediate level.
+        let tier = depth / TieredSmt::TIER_DEPTH;
+        self.depth = tier * TieredSmt::TIER_DEPTH;
+        self
+    }
+}
+
+// CANONICAL WORD
+// ================================================================================================
+
+/// A `[Word]` in canonical representation.
+#[derive(Copy, Clone, Debug, Default, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct CanonicalWord([u64; WORD_SIZE]);
```

Comment:
This struct is relevant because `PartialOrd` won't work out of the box for `Word`. If we don't have a `CanonicalWord`, we will have to

1) Convert `Word` from a type alias to a struct that wraps the array
2) Implement `PartialOrd` for it manually as `Felt` doesn't implement it.

There is a good reason on why `Felt` doesn't implement `PartialOrd`: it cannot know when its inner value is canonical. Whenever you need to order these field elements, you must first convert it to canonical (that is just a `u64`), and then perform the ord.

If we go for approach `2`, we will have, for each iteration of a sort check (that should, normally, be as trivial as possible for a `u64`), convert it's field elements into the canonical form, and then check the order.

We could, alternatively, make the struct `Word` to be exactly what `CanonicalWord` is now, but then we might end up having canonical representations far more than we should, instead of working with `Word`s, when order doesn't matter.